storage/client/internal/parse_rfc3339.cc

```diff
@@ -0,0 +1,207 @@
+                "Invalid format for RFC 3339 timestamp detected while parsing"
+                " the base date and time portion.");
+  }
+  if (date_time_separator != 'T' and date_time_separator != 't') {
+    ReportError(timestamp, "Invalid date-time separator, expected 'T' or 't'.");
+  }
+  if (month <= 0 or month > 12) {
+    ReportError(timestamp, "Out of range month.");
+  }
+  constexpr int MAX_DAYS_IN_MONTH[] = {
+      31,  // January
+      29,  // February (non-leap years checked below)
+      31,  // March
+      30,  // April
+      31,  // May
+      30,  // June
+      31,  // July
+      31,  // August
+      30,  // September
+      31,  // October
+      30,  // November
+      31,  // December
+  };
+  if (day <= 0 or day > MAX_DAYS_IN_MONTH[month - 1]) {
+    ReportError(timestamp, "Out of range month day.");
+  }
+  if (2 == month and day > 28 and not IsLeapYear(year)) {
+    ReportError(timestamp, "Out of range month day.");
+  }
+  if (hours < 0 or hours > 23) {
+    ReportError(timestamp, "Out of range hour.");
+  }
+  if (minutes < 0 or minutes > 59) {
+    ReportError(timestamp, "Out of range minute.");
+  }
+  if (seconds < 0 or seconds > 60) {
+    ReportError(timestamp, "Out of range second.");
+  }
+  // Advance the pointer for all the characters read.
+  buffer += pos;
+
+  std::tm tm{0};
+  tm.tm_year = year - 1900;
+  tm.tm_mon = month - 1;
+  tm.tm_mday = day;
+  tm.tm_hour = hours;
+  tm.tm_min = minutes;
+  tm.tm_sec = seconds;
+  return std::chrono::system_clock::from_time_t(std::mktime(&tm));
+}
+
+std::chrono::system_clock::duration ParseFractionalSeconds(
+    char const*& buffer, std::string const& timestamp) {
+  if (buffer[0] != '.') {
+    return std::chrono::system_clock::duration(0);
+  }
+  ++buffer;
+
+  long fractional_seconds;
+  int pos;
+  auto count = std::sscanf(buffer, "%9ld%n", &fractional_seconds, &pos);
+  if (count != 1) {
+    ReportError(timestamp, "Invalid fractional seconds component.");
+  }
+  // Normalize the fractional seconds to nanoseconds.
+  for (int digits = pos; digits < 9; ++digits) {
+    fractional_seconds *= 10;
+  }
+  // Now skip any other digits ...
+  buffer += pos;
+  while (std::isdigit(buffer[0])) {
+    ++buffer;
+  }
+  return std::chrono::duration_cast<std::chrono::system_clock::duration>(
+      std::chrono::nanoseconds(fractional_seconds));
+}
+
+std::chrono::seconds ParseOffset(char const*& buffer,
+                                 std::string const& timestamp) {
+  if (buffer[0] == '+' or buffer[0] == '-') {
+    bool positive = buffer[0] == '+';
+    ++buffer;
+    // Parse the HH:MM offset.
+    int hours, minutes, pos;
+    auto count = std::sscanf(buffer, "%2d:%2d%n", &hours, &minutes, &pos);
+    constexpr int EXPECTED_OFFSET_WIDTH = 5;
+    constexpr int EXPECTED_OFFSET_FIELDS = 2;
+    if (count != EXPECTED_OFFSET_FIELDS or pos != EXPECTED_OFFSET_WIDTH) {
+      ReportError(timestamp, "Invalid timezone offset, expected [+/-]HH:SS.");
```

Comment:
1) We should keep the bracket syntax consistent throughout our error messages -- if we're using the brackets for regex-style character sets, this should be `[+-]`, not `[+/-]`.

2) `HH:SS` should be `HH:MM`

Reply:
Fixed.